File: Algorithm_113_Path Sum II.cpp
#include <iostream>
#include <vector>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution
{
  public:
    vector<vector<int> > pathSum(TreeNode *root, int sum)
    {
        vector<int> midRes;
        vector<vector<int> > res;
        getPath(root, sum, midRes, res);
        return res;
    }
    void getPath(TreeNode *root, int sum, vector<int> midRes, vector<vector<int> > &res)
    {
        if (!root) return;
        sum -= root->val;
        midRes.push_back(root->val);
        if (!root->left && !root->right && !sum) res.push_back(midRes);
        if (root->left) getPath(root->left, sum, midRes, res);
        if (root->right) getPath(root->right, sum, midRes, res);
    }
};

File: Algorithm_113_Path Sum II.java
// 思路：类似于前序遍历
// 注意函数退出前要把之前添加到cur末尾的root.val删除掉

class Solution {
    private List<List<Integer>> result = new ArrayList<>();
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root == null) {
            return result;
        }
        preOrder(root, sum, new ArrayList<>());
        return result;
    }
    private void preOrder(TreeNode root, int sum, List<Integer> cur) {
        sum -= root.val;
        cur.add(root.val);
        if (root.left == null && root.right == null && sum == 0) {
            result.add(new ArrayList<>(cur));
        } else {
            if (root.left != null) {
                preOrder(root.left, sum, cur);
            }
            if (root.right != null) {
                preOrder(root.right, sum, cur);
            }
        }
        cur.remove(cur.size() - 1);
    }
}